compiler/rustc_hir_analysis/src/check/method/suggest.rs

```diff
@@ -1201,6 +1203,71 @@ impl<'a, 'tcx> FnCtxt<'a, 'tcx> {
         false
     }
 
+    fn suggest_range_for_iter(
+        &self,
+        tcx: TyCtxt<'tcx>,
+        actual: Ty<'tcx>,
+        source: SelfSource<'tcx>,
+        span: Span,
+        item_name: Ident,
+        ty_str: &str,
+    ) -> bool {
+        if let SelfSource::MethodCall(expr) = source {
+            for (_, parent) in tcx.hir().parent_iter(expr.hir_id) {
+                if let Node::Expr(parent_expr) = parent && is_range_literal(parent_expr) {
+                    let span_included = match parent_expr.kind {
+                            hir::ExprKind::Struct(_, eps, _) => eps.iter().any(|ep| ep.expr.span.contains(span)),
+                            // `..=` desugars into `::std::ops::RangeInclusive::new(...)`.
+                            hir::ExprKind::Call(ref func, ..) => func.span.contains(span),
+                            _ => false,
+                    };
+
+                    if !span_included {
+                        continue;
+                    }
+
+                    let range_def_id = self.tcx.lang_items().range_struct().unwrap();
+                    let range_ty = self.tcx.bound_type_of(range_def_id).subst(self.tcx, &[actual.into()]);
+                    let call_expr = self.tcx.hir().expect_expr(self.tcx.hir().get_parent_node(expr.hir_id));
```

Comment:
Why are we certain that `expr`'s parent is an expr?
Should we just exit the loop the first time we do not see an expr?

Reply:
we are sure about this because this check: https://github.com/rust-lang/rust/blob/348d08282afc03c159cdf2adceacedf37ac13971/compiler/rustc_hir_analysis/src/check/method/suggest.rs#L1217

But, this `call_expr` is not necessary now, I have removed it in my local branch, will push it later.